### PR TITLE
Use the humantime_to_duration crate

### PR DESCRIPTION
**Pull request by sylvestre**

### CODE REVIEW

**Comment by Joining7943** [comment]
Sorry, I wrote this into a closed pr: I think `fundu` could be capable of parsing such strings into durations.

**Comment by sylvestre** [comment]
@Joining7943 TIL, thanks :)
However, fundu manages way more formats than we probably want (and that GNU does)

File: src/uu/touch/src/touch.rs [hunk 1]
```diff
@@ -83,32 +83,35 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+                    let ref_mtime_secs = mtime.unix_seconds();
+                    let ref_mtime_nanos = mtime.nanoseconds();
+                    let mtime = FileTime::from_unix_time(
+                        ref_mtime_secs + seconds,
+                        ref_mtime_nanos + nanos as u32,
+                    );
+
+                    (atime, mtime)
+                }
+                Err(..) => {
```
Comment (tertsdiepraam) [inline src/uu/touch/src/touch.rs:111]: Happy to accept this as a first version, but the date parsing would need to be part of the humantime crate and ignoring the error here is not quite right. We could open an issue for this once this PR is merged.

**Review by tertsdiepraam — APPROVED** [review]
Nice! Now we can move the `date` util over too and actually work on a good implementation :)

**Comment by sylvestre** [comment]
> Nice! Now we can move the date util over too and actually work on a good implementation :)

Sounds good :)

I was also thinking that we could use this kind of feature in other commands

for example
ls --older-than "3 days ago" 

or in most of the command which relies on date

**Comment by tertsdiepraam** [comment]
Good idea, would be a nice extension to GNU!

File: src/uu/touch/src/touch.rs [hunk 2]
```diff
@@ -83,32 +83,35 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
     ) {
         (Some(reference), Some(date)) => {
             let (atime, mtime) = stat(Path::new(reference), !matches.get_flag(options::NO_DEREF))?;
-            if let Some(offset) = parse_relative_time(date) {
-                let mut seconds = offset.whole_seconds();
-                let mut nanos = offset.subsec_nanoseconds();
-                if nanos < 0 {
-                    nanos += 1_000_000_000;
-                    seconds -= 1;
+            match humantime_to_duration::from_str(date) {
```
Comment (cakebaker) [inline src/uu/touch/src/touch.rs:86]: Why do you replace `if let` with `match`?

File: tests/by-util/test_touch.rs [hunk 3]
```diff
@@ -608,7 +608,17 @@ fn test_touch_set_date_relative_smoke() {
...
+    let times = [
+        "yesterday",
+        "tomorrow",
+        "now",
+        "2 seconds",
+        "second",
+        "2 years 1 week",
+        "2 days ago",
+        "2 months and 1 second",
+        "fortnight",
```
Comment (cakebaker) [inline tests/by-util/test_touch.rs:620]: `second` and `fortnight` are already in the other `times` array a few lines above and so are tested twice.

**Comment by Joining7943** [comment]
> However, fundu manages way more formats than we probably want (and that GNU does)

The format of `fundu` can be adjusted to the one that GNU uses as far as I've seen. The only thing that's currently not realizable is `ago`. But as long as the `ago` keyword can only end a relative time this is not really a big issue. Note I'm talking about the currently unfinished version `0.5.1` of `fundu` currently living on the `develop` branch. I didn't had much time for a new release, so `0.5.1` is pending since a while.

**Comment by sylvestre** [comment]
oh, you are the upstream of fundu :)

**Comment by Joining7943** [comment]
yes :) I don't want to make false promises and I would have to try myself first but I think gnu relative times might be realizable by `fundu`. They look pretty similar to `systemd` time spans and the goal of `0.5.1` was to be able to build a systemd time span parser. If `fundu` can't be used for gnu relative times now, I'm confident it could in `v0.5.2`...

**Comment by tertsdiepraam** [comment]
Just keep in mind that all of this is the goal here: https://www.gnu.org/software/coreutils/manual/html_node/Date-input-formats.html

**Comment by Joining7943** [comment]
> Just keep in mind that all of this is the goal here: https://www.gnu.org/software/coreutils/manual/html_node/Date-input-formats.html It's a lot to add and serves a different purpose.

I have no intention to support the whole gnu date format in `fundu`. But that relative time parsing part (https://www.gnu.org/software/coreutils/manual/html_node/Relative-items-in-date-strings.html ) shouldn't be a problem.

**Comment by sylvestre** [comment]
I will land this PR because it is already providing the feature. we can always iterate later

**Pull request merged**